### PR TITLE
Do not generate buffers just to close them.

### LINKED ISSUE #50: Clean up exceptions in OBSFile.__del__
bunch of failures there that should be fixed up by checking whether we've actually downloaded the buffer (and not attempting to read in the buffer on `__del__`).

Basically just a `if 'buffer' in self.__dict__` kinda thing

### PR DESCRIPTION
**Pull request by jtratner**
Previously we were creating buffers in the ``close()`` method in order
to call ``close()`` on them and/or causing uploads in the close method
in an attempt to upload an empty file. For strict compatibility with
POSIX, we probably should create an empty file if we open but do
not close an object, but this means we can potentially generate an error
in a close method in an edge case that isn't that fun.

This is not an API-breaking change, because previously we checked for
buffer.tell() in flush, so no empty files are created.

Closes #50.

### CODE REVIEW

**Comment by jtratner** [comment]
Back to you @pkaleta - added requested test cases and such

File: stor/obs.py [hunk 1]
```diff
@@ -256,17 +255,27 @@ class OBSFile(object):
...
 
         with Path('swift://tenant/container/object').open(mode='r') as obj:
             obj.write('Hello world!')
+
+    .. note::
+        Unlike python file objects, OBSFile does not create an empty sentinel file if you
+        call open and then do not close it.
     """
     closed = False
+    _buffer_created = False
```
[the remark this answers is not in the document]
Comment (jtratner) [reply stor/obs.py]: it was the only way I could think of to test the behavior that buffers weren't being created on close, while still freeing the underlying object.

I guess if I just don't delete the reference to the buffer and let that be garbage collected with closed file, that'd be fine.

File: stor/obs.py [hunk 2]
```diff
@@ -256,17 +255,27 @@ class OBSFile(object):
...
     _READ_MODES = ('r', 'rb')
     _WRITE_MODES = ('w', 'wb')
```
[the remark this answers is not in the document]
Comment (jtratner) [reply stor/obs.py]: we didn't previously support it, I'd rather add support in a follow up PR

File: stor/obs.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jtratner) [reply stor/obs.py]: I'll make that change

File: stor/obs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jtratner) [reply stor/obs.py]: I *think* you could add support by (weirdly) just adding `w+` to both `READ_MODES` and `WRITE_MODES`

File: stor/obs.py [hunk 3]
```diff
@@ -351,18 +366,24 @@ def name(self):
...
         """Flushes the write buffer to the OBS path (if it exists)"""
         if self.mode not in self._WRITE_MODES:
             raise TypeError("File must be in modes %s to 'flush'" %
                             (self._WRITE_MODES,))
-        if self._buffer.tell():
+        # NOTE: this helps ensure that only non-zero objects are uploaded with open.
+        # Otherwise you have weird behavior where calling open().tell() will cause an empty object
+        # to be created on OBS on exit. Instead, philosophy is "if buffer has data, we'll upload it
+        # on close"
+        if self.__buffer and self._buffer.tell():
```
[the remark this answers is not in the document]
Comment (jtratner) [reply stor/obs.py]: this is an excellent point.

File: stor/obs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pkaleta) [reply stor/obs.py]: another PR makes sense

File: stor/tests/shared_obs.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    """Tests that only rely on read_object() on underlying path class to be called, so we can just
+    parametrize on drive and path"""
+    drive = None
+    path_class = None
+
+    def test_works_with_gzip(self):
+        gzip_path = stor.join(stor.dirname(__file__),
+                              'file_data', 's_3_2126.bcl.gz')
+        text = stor.open(gzip_path, 'rb').read()
+        with mock.patch.object(self.path_class, 'read_object', autospec=True) as read_mock:
```
[the remark this answers is not in the document]
Comment (jtratner) [reply stor/tests/shared_obs.py]: done

File: stor/tests/shared_obs.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+            read_mock.return_value = text
+            fileobj = stor.open('{drive}A/C/s_3_2126.bcl.gz'.format(drive=self.drive), 'rb')
```
[the remark this answers is not in the document]
Comment (jtratner) [reply stor/tests/shared_obs.py]: added that

[the remark this answers is not in the document]
Comment (jtratner) [reply stor/tests/shared_obs.py]: good catch - added

**Comment by jtratner** [comment]
Decided that the changes here were backwards-compatible enough that I'm going to have this merge into the v2.0 branch.

File: stor/tests/shared_obs.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    def test_invalid_open(self):
+        pth = stor.join(self.drive, 'B/C/D')
+        with self.assertRaisesRegexp(ValueError, 'mode'):
+            # keep reference here
+            f = stor.open(pth, 'invalid')  # nopep8
+            assert False, 'should error before this'  # pragma: no cover
+
+        with self.assertRaisesRegexp(ValueError, 'mode'):
+            with stor.open(pth, 'invalid'):
+                assert False, 'should error before this'  # pragma: no cover
```
Comment (pkaleta) [inline stor/tests/shared_obs.py]: 👍

**Pull request merged**